README.md
# NAME

Redis::Sweeten - ...

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.16+-blue.svg" alt="Requires Perl 5.16+" />
    <a href="https://travis-ci.org//"><img src="https://api.travis-ci.org//.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Redis-Sweeten-0.0101"><img src="https://badgedepot.code301.com/badge/kwalitee/Redis-Sweeten/0.0101" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Redis-Sweeten%200.0101"><img src="https://badgedepot.code301.com/badge/cpantesters/Redis-Sweeten/0.0101" alt="CPAN Testers result" /></a>
    </p>
</div>

# VERSION

Version 0.0101, released 2016-03-22.

# SYNOPSIS

    use Redis::Fast::Sweeten;

# DESCRIPTION

Redis::Fast::Sweeten is ...

# SEE ALSO

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Redis-Sweeten

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0101 2016-03-22T12:12:14Z UTC
 - Initial release

 [Stats & Dependency Changes]
 - (run req) + Class::Method::Modifiers (any)
 - (run req) + Mojolicious (any)
 - (run req) + Moose::Role (any)
 - (run req) + MooseX::AttributeShortcuts 0.028
 - (run req) + Redis::Fast (any)
 - (run req) + Try::Tiny 0.24
 - (run req) + Types::Standard 1.000005
 - (run req) + experimental (any)
 - (run req) + namespace::autoclean 0.22
 - (run req) + perl 5.016000
 - (test req) + ExtUtils::MakeMaker (any)
 - (test req) + File::Spec (any)
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (test req) + Test::More 0.96
 - (test rec) + CPAN::Meta 2.120900
 - (conf req) + ExtUtils::MakeMaker (any)
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
 - (dev req) + Dist::Zilla::Plugin::CheckChangesHasContent 0.008
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::ConfirmRelease (any)
 - (dev req) + Dist::Zilla::Plugin::CopyFilesFromBuild 0.151680
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::ExecDir (any)
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::Git::Check (any)
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors 0.017
 - (dev req) + Dist::Zilla::Plugin::Git::GatherDir (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Push (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Tag (any)
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::License (any)
 - (dev req) + Dist::Zilla::Plugin::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::Manifest (any)
 - (dev req) + Dist::Zilla::Plugin::ManifestSkip (any)
 - (dev req) + Dist::Zilla::Plugin::MetaJSON (any)
 - (dev req) + Dist::Zilla::Plugin::MetaNoIndex (any)
 - (dev req) + Dist::Zilla::Plugin::MetaProvides::Package 2.003001
 - (dev req) + Dist::Zilla::Plugin::MetaYAML (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::PodSyntaxTests (any)
 - (dev req) + Dist::Zilla::Plugin::PodWeaver 4.006
 - (dev req) + Dist::Zilla::Plugin::Prereqs (any)
 - (dev req) + Dist::Zilla::Plugin::Readme (any)
 - (dev req) + Dist::Zilla::Plugin::ReadmeAnyFromPod 0.150250
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::RunExtraTests 0.028
 - (dev req) + Dist::Zilla::Plugin::ShareDir (any)
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.054
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::ReportPrereqs 0.021
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Dist::Zilla::Plugin::TestRelease (any)
 - (dev req) + Dist::Zilla::Plugin::TravisYML 1.15
 - (dev req) + Dist::Zilla::Plugin::UploadToStratopan 0.012
 - (dev req) + Pod::Elemental::Transformer::List (any)
 - (dev req) + Pod::Weaver::Plugin::SingleEncoding (any)
 - (dev req) + Pod::Weaver::Plugin::Transformer (any)
 - (dev req) + Pod::Weaver::PluginBundle::CorePrep (any)
 - (dev req) + Pod::Weaver::Section::Authors (any)
 - (dev req) + Pod::Weaver::Section::Badges 0.0401
 - (dev req) + Pod::Weaver::Section::Collect (any)
 - (dev req) + Pod::Weaver::Section::Generic (any)
 - (dev req) + Pod::Weaver::Section::Leftovers (any)
 - (dev req) + Pod::Weaver::Section::Legal (any)
 - (dev req) + Pod::Weaver::Section::Name (any)
 - (dev req) + Pod::Weaver::Section::Region (any)
 - (dev req) + Pod::Weaver::Section::Version (any)
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Pod 1.41
 - (dev req) + Test::Version 1
 - (dev req) + Test::Warnings 0.026
 - (dev sug) + Dist::Iller 0.1408
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0309


lib/Redis/Sweeten.pm
use 5.20.0;
use strict;
use warnings;

package Redis::Sweeten;

our $VERSION = '0.0102';
# ABSTRACT: ...

    use Moose::Role;
    use MooseX::AttributeShortcuts;
    use Types::Standard -types;
    use Redis::Fast;
    use Encode;
    use Try::Tiny;
    use namespace::autoclean;
    use experimental qw/postderef signatures/;

    has redis => (
        is => 'rw',
        lazy => 1,
        builder => 1,
        clearer => 1,
        predicate => 1,
    );
    has not_connected_since => (
        is => 'rw',
        isa => Int,
        init_arg => 1,
        clearer => 1,
        predicate => 1,
    );
    has attempt_reconnect_every => (
        is => 'ro',
        isa => Int,
        default => 15,
    );

    requires qw/server prefix default_expiry/;

    sub connection_name($self) { $self->prefix }

    sub _build_redis($self) {
        my $redis;
        try {
            $redis = Redis::Fast->new(server => $self->server, name => $self->connection_name);
        }
        catch { };
        return $redis;
    };

    around [qw/get get_utf8 set setex del/] => sub ($next, $self, $key, @args) {
        if(time - $self->has_not_connected_since > $self->attempt_reconnect_every) {

            $self->not_connected_since(time);

            if(!$self->has_redis) {
                $self->redis($self->_build_redis);

                if(defined $self->redis && $self->redis->ping) {
                    $self->clear_not_connected_since;
                }
            }
        }
        $key = join '.' => ($self->prefix, $key);
        try {
            $self->$next($key, @args);
        }
        catch {
            $self->clear_redis;
            $self->not_connected_since(time);
            return undef;
        };
    };

    sub get_utf8($self, $key) {
        return decode('UTF-8', $self->redis->get($key));
    }
    sub get($self, $key) {
        return $self->redis->get($key);
    }
    sub set($self, $key, $value) {
        return $self->redis->set($key, $value);
    }

    # @data can be 1 or 2: the last item is always the value.
    # expiration time is optional, but is the first item (if there are two)
    # this is keep the order of arguments for redis setex
    sub setex($self, $key, @data) {
        my $value = pop @data;
        my $expiry = pop @data || $self->default_expiry;

        return $self->redis->setex($key, $expiry, $value);
    }

    sub del($self, $key) {
        return $self->redis->del($key);
    }


1;

__END__

=pod

=head1 SYNOPSIS

    use Redis::Fast::Sweeten;

=head1 DESCRIPTION

Redis::Fast::Sweeten is ...

=head1 SEE ALSO

=cut


lib/Mojolicious/Plugin/RedisHandler.pm
use 5.20.0;
use warnings;

package Mojolicious::Plugin::RedisHandler;

our $VERSION = '0.0102';
# ABSTRACT: ...

    use Mojo::Base 'Mojolicious::Plugin';
    use Class::Method::Modifiers;
    use Types::Standard -types;
    use Redis::Fast;
    use Encode;
    use MIME::Base64 qw/encode_base64 decode_base64/;
    use Try::Tiny;
    use namespace::autoclean;
    use experimental qw/postderef signatures/;

    has not_connected_since => undef;
    has attempt_reconnect_every => 15;
    has prefix => undef;
    has separator => ':';
    has database => 0;
    has connection_name => undef;
    has default_expiry => 900;
    has server => undef;
    has actual => undef;

    sub register($self, $app, $conf) {
        if(!exists $conf->{'server'}) {
            die "M::P::RedisHandler: mandatory configuration 'server' is missing";
        }
        $self->server($conf->{'server'});

        if(exists $conf->{'attempt_reconnect_every'}) {
            if($conf->{'attempt_reconnect_every'} =~ m{\D} || $conf->{'attempt_reconnect_every'} < 0) {
                die "M::P::RedisHandler: mandatory configuration 'attempt_reconnect_every' must be non-negative integer (was $conf->{'attempt_reconnect_every'})";
            }
            $self->attempt_reconnect_every($conf->{'attempt_reconnect_every'});
        }

        if(exists $conf->{'prefix'}) {
            $self->prefix($conf->{'prefix'});
        }
        if(exists $conf->{'connection_name'}) {
            $self->connection_name($conf->{'connection_name'});
        }
        if(exists $conf->{'default_expiry'}) {
            $self->default_expiry($conf->{'default_expiry'});
        }
        if(exists $conf->{'separator'}) {
            $self->separator($conf->{'separator'});
        }
        $self->database($conf->{'database'});

        $self->actual($self->_build_redis);

        my $handler_name = $conf->{'helper'} || 'redis';
        $app->helper($handler_name => sub { $self });

        return $self;
    }

    sub _build_redis($self) {
        my $redis;
        try {
            $redis = Redis::Fast->new(server => $self->server, name => $self->connection_name);
        }
        catch { };
        return $redis;
    };

    around [qw/get get_utf8 set setex del/] => sub ($next, $self, $key, @args) {
        warn ">>$key";
        if(!defined $self->actual && defined $self->not_connected_since && time - $self->not_connected_since > $self->attempt_reconnect_every) {
            $self->actual($self->_build_redis);

            if(defined $self->actual && $self->actual->ping) {
                $self->not_connected_since(undef);
            }
            else {
                $self->not_connected_since(time);
            }
        }

        $key = encode_base64 join ('.' => (defined $self->prefix ? $self->prefix : (), $key)), '';
        try {
            return $self->$next($key, @args);
        }
        catch {
            warn 'catched!' . $_;
            $self->not_connected_since(time);
            $self->actual(undef);
            return undef;
        };
    };

    sub get_utf8($self, $key) {
        return decode('UTF-8', $self->actual->get($key));
    }
    sub get($self, $key) {
        return $self->actual->get($key);
    }
    sub set($self, $key, $value) {
        return $self->actual->set($key, $value);
    }

    # @data can be 1 or 2: the last item is always the value.
    # expiration time is optional, but is the first item (if there are two)
    # this is keep the order of arguments for redis setex
    sub setex($self, $key, @data) {
        my $value = pop @data;
        my $expiry = pop @data || $self->default_expiry;

        return $self->actual->setex($key, $expiry, $value);
    }

    sub del($self, $key) {
        return $self->actual->del($key);
    }



1;

__END__

=pod

=head1 SYNOPSIS

    use Mojolicious::Plugin::RedisHandler;

=head1 DESCRIPTION

Mojolicious::Plugin::RedisHandler is ...

=head1 SEE ALSO

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.016000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "...",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Redis-Sweeten",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.016000",
  "NAME" => "Redis::Sweeten",
  "PREREQ_PM" => {
    "Class::Method::Modifiers" => 0,
    "Mojolicious" => 0,
    "Moose::Role" => 0,
    "MooseX::AttributeShortcuts" => "0.028",
    "Redis::Fast" => 0,
    "Try::Tiny" => "0.24",
    "Types::Standard" => "1.000005",
    "experimental" => 0,
    "namespace::autoclean" => "0.22"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0102",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Class::Method::Modifiers" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Mojolicious" => 0,
  "Moose::Role" => 0,
  "MooseX::AttributeShortcuts" => "0.028",
  "Redis::Fast" => 0,
  "Test::More" => "0.96",
  "Try::Tiny" => "0.24",
  "Types::Standard" => "1.000005",
  "experimental" => 0,
  "namespace::autoclean" => "0.22"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
